CS483/CS483/Kartaclysm/StateMachine/Gameplay/StatePaused.cpp

```diff
@@ -0,0 +1,166 @@
+
+	if (m_pPauseDelegate != nullptr)
+	{
+		// If this is the first frame suspended: remove listener
+		HeatStroke::EventManager::Instance()->RemoveListener("Pause", m_pPauseDelegate);
+		delete m_pPauseDelegate;
+		m_pPauseDelegate = nullptr;
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:		Unsuspend
+// Parameter:	const int p_iPrevState - index of previous state popped
+// 
+// Called when this state is popped back to top of stack.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Unsuspend(const int p_iPrevState)
+{
+	m_bSuspended = false;
+	
+	if (m_pPauseDelegate == nullptr)
+	{
+		m_pPauseDelegate = new std::function<void(const HeatStroke::Event*)>(std::bind(&StatePaused::PauseGame, this, std::placeholders::_1));
+		HeatStroke::EventManager::Instance()->AddListener("Pause", m_pPauseDelegate);
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    Update
+// Parameter: const float p_fDelta
+// 
+// Called each from when this state is active.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Update(const float p_fDelta)
+{
+	// Do not update when suspended
+	if (!m_bSuspended)
+	{
+		m_pGameObjectManager->Update(p_fDelta);
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    PreRender
+// 
+// Called before rendering occurs.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::PreRender()
+{
+	// Render even when suspended
+	m_pGameObjectManager->PreRender();
+}
+
+//------------------------------------------------------------------------------
+// Method:    Exit
+// 
+// Called when this state is popped off the stack.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::Exit()
+{
+	m_bSuspended = false;
+
+	if (m_pPauseDelegate != nullptr)
+	{
+		HeatStroke::EventManager::Instance()->RemoveListener("Pause", m_pPauseDelegate);
+		delete m_pPauseDelegate;
+		m_pPauseDelegate = nullptr;
+	}
+
+	if (m_pGameObjectManager != nullptr)
+	{
+		m_pGameObjectManager->DestroyAllGameObjects();
+		delete m_pGameObjectManager;
+		m_pGameObjectManager = nullptr;
+	}
+}
+
+//------------------------------------------------------------------------------
+// Method:    PauseGame
+// Parameter: const HeatStroke::Event* p_pEvent - Event that triggers when a player pauses
+// 
+// Unpause the game by popping the Pause State.
+//------------------------------------------------------------------------------
+void Kartaclysm::StatePaused::PauseGame(const HeatStroke::Event* p_pEvent)
+{
+	// Get the player who paused the game
+	int iPlayer = 0;
+	p_pEvent->GetOptionalIntParameter("Player", iPlayer, iPlayer);
+
+	// Pop current state
+	printf("Unpausing\n");
```

Comment:
It's not just a problem with pausing/unpausing, it's also a problem with abilities at well. It's something that should be handled at the Input level, especially since they already keep track of the previous frame's inputs.

It's a simple method to add to the Input Buffers, ButtonDownOnce() vs ButtonDownContinuous() as examples, maybe ButtonReleaseOnce() if there's a need. I slotted this fix into the Backlog for the next sprint, or at least until I see how long the HUD is going to take me for this sprint.